crates/astria-sequencer/src/accounts/action.rs

```diff
@@ -46,21 +68,56 @@ impl ActionHandler for TransferAction {
             amount = self.amount,
         )
     )]
-    async fn execute<S: StateWriteExt>(&self, state: &mut S, from: Address) -> Result<()> {
+    async fn execute<S: StateWriteExt>(
+        &self,
+        state: &mut S,
+        from: Address,
+        fee_asset_id: &asset::Id,
+    ) -> Result<()> {
+        let transfer_asset_id = self.asset_id;
+
         let from_balance = state
-            .get_account_balance(from)
+            .get_account_balance(from, &transfer_asset_id)
             .await
             .context("failed getting `from` account balance")?;
         let to_balance = state
-            .get_account_balance(self.to)
+            .get_account_balance(self.to, &transfer_asset_id)
             .await
             .context("failed getting `to` account balance")?;
-        state
-            .put_account_balance(from, from_balance - (self.amount + TRANSFER_FEE))
-            .context("failed updating `from` account balance")?;
-        state
-            .put_account_balance(self.to, to_balance + self.amount)
-            .context("failed updating `to` account balance")?;
+
+        // if fee payment asset is same asset as transfer asset, deduct fee
+        // from same balance as asset transferred
+        if &transfer_asset_id == fee_asset_id {
+            state
+                .put_account_balance(
+                    from,
+                    &transfer_asset_id,
+                    from_balance - (self.amount + TRANSFER_FEE),
```

Comment:
As above, can we use checked math for all of this? I.e. `checked_sub` + `checked_add`.